packages/excalidraw/components/App.tsx

```diff
@@ -6349,15 +6352,76 @@ class App extends React.Component<AppProps, AppState> {
   private handleCanvasPointerUp = (
     event: React.PointerEvent<HTMLCanvasElement>,
   ) => {
-    this.removePointer(event);
-    this.lastPointerUpEvent = event;
+    this.removePointer(event); // always remove pointer on pointerUp
+    this.lastPointerUpEvent = event; // save last pointer up event
 
     const scenePointer = viewportCoordsToSceneCoords(
       { clientX: event.clientX, clientY: event.clientY },
       this.state,
     );
     const clicklength =
-      event.timeStamp - (this.lastPointerDownEvent?.timeStamp ?? 0);
+      event.timeStamp - (this.lastPointerDownEvent?.timeStamp ?? 0); // this is to distinguish between click and drag; a click is less than 300ms, a drag is more than 300ms; 300ms comes from the default value of the clicklength in the browser
+
+    // Ensure lastPointerDownEvent is not defined before accessing its properties
+    if (!this.lastPointerDownEvent) {
+      return; // return if lastPointerDownEvent is not defined
+    }
+
+    // Variable to track the initial pointer down position => this happens right before the pointer up event to compare with the pointer up position
+    const pointerDownPosition = {
+      x: this.lastPointerDownEvent?.clientX, // x position of the pointer down event
+      y: this.lastPointerDownEvent?.clientY, // y position of the pointer down event; clientY is a read-only property of the MouseEvent interface that returns the vertical coordinate within the application's viewport at which the event occurred (as opposed to the coordinate within the page).
+    };
+
+    // Calculate the distance moved during the inital pointer down position to the pointer up position
+    const distanceMoved = Math.sqrt(
+      Math.pow(event.clientX - pointerDownPosition.x, 2) +
+        Math.pow(event.clientY - pointerDownPosition.y, 2),
+    );
+
+    const addElement = (element: any /* : Partial<ExcalidrawElement> */) => {
+      // TODO: here we need the right ExcalidrawElement type
+      const newElement = {
+        ...element,
+        // const id = ... ?
+        // isDeleted: false ?
+        // other properties as needed
+      };
+    };
+
+    // if the distance is less than 5px and the click length is less than 300ms, then it is a click
+    const dragTreshold = 5;
+    if (distanceMoved < dragTreshold && clicklength < 300) {
+      // check the selected tool for shape creation
+      if (
+        [
+          "rectangle",
+          "ellipse",
+          "diamond",
+          "arrow",
+          "line",
+          "freedraw",
+        ].includes(this.state.activeTool.type)
+      ) {
+        const newElement = {
```

Comment:
Sure, that´s good to know! 

I tried it with this approach now: 

```
if (
          this.state.activeTool.type === "line" ||
          this.state.activeTool.type === "arrow"
        ) {
          // Provide points for line and arrow tools
          newElement.points = [
            pointFrom(0, 0), // Starting point
            pointFrom(0, 10), // End point
          ];
        }
```

The problem with this is, that the Property 'points' does not exist on type '{ type: ToolType | "custom"; width: number; height: number; x: number; y: number; }'

Am I allowed to add this property to the Tool interface and if so, where is it located?

By updating the type definition to include the points property and ensuring that newElements conform to this type, I could resolve the TS Error, but as I don´t know the whole codebase just as good as you do, I´d rather take your advice. 

Thank you!

Reply:
First of all, the element `points` is a read-only property, so ideally you'd add this when you call `newElement({ ..., points: [....]})`. 

Given that not all types of elements have a `points` property, this approach will also guide you to solve this problem you mention above, because when you provide parameters to newElement(...) you are forced to define the `type` of the element being created, which will properly validate when `points` can be added to a new element.